base/sparse/sparsematrix.jl

```diff
@@ -2252,6 +2252,8 @@ function setindex!{Tv}(A::SparseMatrixCSC{Tv}, x::Number,
     if isempty(I) || isempty(J); return A; end
     if !issorted(I); I = sort(I); end
     if !issorted(J); J = sort(J); end
+    if !allunique(I); I = unique(I); end
```

Comment:
But we can only use `sort!` in the case where there are duplicates then (such that we have `I = unique(I)` making a copy and then sort in place). The input vectors should not be mutated so `unique` or `sort` need to make a copy. I would think it is more common with unsorted vectors than vectors with duplicates, right? In that case, wouldn't this be better:

```julia
if !issorted(I)
    if !allunique(I)
        I = sort!(unique(I))
    else # unique but not sorted
        I = sort(I)
    end
elseif !allunique(I) # sorted
    I = unique(I)
end
```